src/gpl2/include/gpl2/DgReplace.h

```diff
@@ -0,0 +1,212 @@
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+namespace odb {
+class dbDatabase;
+class dbInst;
+}  // namespace odb
+namespace sta {
+class dbSta;
+class dbNetwork;
+}  // namespace sta
+
+namespace grt {
+class GlobalRouter;
+}
+
+namespace rsz {
+class Resizer;
+}
+
+namespace utl {
+class Logger;
+}
+
+namespace gpl2 {
+
+class PlacerBaseCommon;
+class PlacerBase;
+class RouteBase;
+class TimingBase;
+class InitialPlace;
+class NesterovPlace;
+
+class DgReplace
+{
+ public:
+  DgReplace();
+  ~DgReplace();
+
+  void init(sta::dbNetwork* network,
+            odb::dbDatabase* odb,
+            rsz::Resizer* resizer,
+            grt::GlobalRouter* router,
+            utl::Logger* logger);
+  void reset();
+
+  // The three main functions
+  void doInitialPlace();
+  int doNesterovPlace(int start_iter = 0);
+
+  // Initial Place param settings
+  void setInitialPlaceMaxIter(int iter);
+  void setInitialPlaceMinDiffLength(int length);
+  void setInitialPlaceMaxSolverIter(int iter);
+  void setInitialPlaceMaxFanout(int fanout);
+  void setInitialPlaceNetWeightScale(float scale);
+
+  void setNesterovPlaceMaxIter(int iter);
+
+  void setBinGridCnt(int binGridCntX, int binGridCntY);
+
+  void setTargetDensity(float density);
+  void setUniformTargetDensityMode(bool mode);
+  void setTargetOverflow(float overflow);
+  void setInitDensityPenalityFactor(float penaltyFactor);
+  void setInitWireLengthCoef(float coef);
+  void setMinPhiCoef(float minPhiCoef);
+  void setMaxPhiCoef(float maxPhiCoef);
+
+  float getUniformTargetDensity();
+
+  // HPWL: half-parameter wire length.
+  void setReferenceHpwl(float deltaHpwl);
+  void setHaloWidth(float haloWidth);
+  void setVirtualIter(int iter);
+  void setNumHops(int hops);
+
+  // temp funcs; OpenDB should have these values.
+  void setPadLeft(int padding);
+  void setPadRight(int padding);
+  void setTimingDrivenMode(bool mode);
+  void setSkipIoMode(bool mode);
+
+  void setRoutabilityDrivenMode(bool mode);
+  void setRoutabilityCheckOverflow(float overflow);
+  void setRoutabilityMaxDensity(float density);
+
+  void setRoutabilityMaxBloatIter(int iter);
+  void setRoutabilityMaxInflationIter(int iter);
+
+  void setRoutabilityTargetRcMetric(float rc);
+  void setRoutabilityInflationRatioCoef(float ratio);
+  void setRoutabilityMaxInflationRatio(float ratio);
+
+  void setRoutabilityRcCoefficients(float k1, float k2, float k3, float k4);
+
+  void addTimingNetWeightOverflow(int overflow);
+  void setTimingNetWeightMax(float max);
+
+  void setDatapathFlag(bool flag);
+  void setDataflowFlag(bool flag);
+  void setClusterConstraintFlag(bool flag);
+
+ private:
+  bool initNesterovPlace();
+
+  sta::dbNetwork* network_;
+  odb::dbDatabase* db_;
+  rsz::Resizer* rs_;
+  grt::GlobalRouter* fr_;
+  utl::Logger* log_;
+
+  // We should only have one placerBaseCommon, timingBase and routeBase
+  // But we need multiple placerBases to handle fences and multiple domains
```

Comment:
This comment seems to have become separate from its context (move down 4 lines)

Reply:
Done